### PR TITLE
Document the ClipboardUnsanitizedFormats parameter to the Clipboard.read() method

### PR DESCRIPTION
**Pull request by captainbrosset**
### Description

Recently, a new optional parameter was added to the `read()` method of the Clipboard interface. This parameter allows developers to specify the mime-types for which they do not want the browser to sanitize the content. In certain cases, browsers may sanitize the clipboard content when reading it. For example, when reading HTML code, chrome inlines CSS styles and removes script tags to avoid pasting malicious content.

In some situations, this can be a problem, for example when building an HTML editor.

To work around this, the new optional parameter was added.

### Motivation

The spec changed, and Chromium browsers now support the parameter. So it's time to update the documentation.

### Additional details

Apple and Mozilla might not actually ever implement this. Even though it's standardized and implemented in chromium, not all browsers are going to agree with it. Mozilla and Apple have said they want the behavior to be consistent whether you use the new async clipboard API or the old one, and that the HTML should either always be sanitized, or never, but that having an option makes no sense to them.

Mozilla: https://github.com/mozilla/standards-positions/issues/769#issuecomment-1852250385
Apple: I couldn't find a standards position for webkit.

### See also

https://developer.chrome.com/docs/web-platform/unsanitized-html-async-clipboard
https://chromium-review.googlesource.com/c/chromium/src/+/3828553

### CODE REVIEW

File: files/en-us/web/api/clipboard/read/index.md [hunk 1]
```diff
@@ -17,11 +17,15 @@ Browsers commonly support reading text, HTML, and PNG image data — see [browse
...
 read()
+read(formats)
 ```
 
 ### Parameters
 
-None.
+- `formats` {{optional_inline}}
+  - : An optional object with the following properties:
+    - `unsanitized`
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:27]:
```suggestion
    - `unsanitized` {{optional_inline}}
```

the spec allow this method to pass an empty object without `unsanitized` option, see also the spec https://w3c.github.io/clipboard-apis/

File: files/en-us/web/api/clipboard/read/index.md [hunk 2]
```diff
@@ -240,6 +244,107 @@ Notes:
 - If prompted, you will need to grant permission in order to paste the image.
 - This may not work on chromium browsers as the sample frame is not granted the [Permissions-Policy](/en-US/docs/Web/HTTP/Headers/Permissions-Policy) `clipboard-read` and `clipboard-write` permissions ([required by Chromium browsers](/en-US/docs/Web/API/Clipboard_API#security_considerations)).
 
+### Reading unsanitized HTML from the clipboard
+
+This example uses the `read(formats)` method to read HTML data from the clipboard and get the code in its original form, without the browser sanitizing it first.
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:249]:
```suggestion
This example uses the `read()` method to read HTML data from the clipboard and get the code in its original form, without the browser sanitizing it first.
```

I think the parameter here is not necessary

File: files/en-us/web/api/clipboard/read/index.md [hunk 3]
```diff
@@ -240,6 +244,107 @@ Notes:
...
+#### HTML
+
+```html
+<textarea id="source" rows="5">
+<style>h1 {color: red;} p {color: blue;}</style><h1>Hello world!</h1><p>This is a test.</p><script>alert('Hello world!');</script></textarea
+>
+<button id="copy">Copy HTML</button>
+<button id="paste_normal">Paste HTML</button>
+<button id="paste_unsanitized">Paste unsanitized HTML</button>
+<textarea id="destination" rows="5"></textarea>
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:262]: pretty the html code is better

image

File: files/en-us/web/api/clipboard/read/index.md [hunk 4]
```diff
@@ -240,6 +244,107 @@ Notes:
...
+    destinationTextarea.value = `Clipboard read failed: ${e}`;
+  }
+});
+```
+
+#### Result
+
+First click the "Copy HTML" button to write the HTML code from the first textarea to the clipboard. Then either click the "Paste HTML" button or the "Paste unsanitized HTML" button to paste the sanitized or unsanitized HTML code into the second textarea.
+
+{{EmbedLiveSample("Reading unsanitized HTML from the clipboard", "100%", "220")}}
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:346]: when running the live example, following errors has been reported:

image

File: files/en-us/web/api/clipboard/read/index.md [hunk 5]
```diff
@@ -240,6 +244,107 @@ Notes:
...
+pasteButton.addEventListener("click", async () => {
+  try {
+    const clipboardContents = await navigator.clipboard.read();
+    const html = await getHTMLFromClipboardContents(clipboardContents);
+    if (html) {
+      destinationTextarea.value = html;
+    }
+  } catch (error) {
+    destinationTextarea.value = `Clipboard read failed: ${e}`;
+  }
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:324]:
```suggestion
  } catch (error) {
    destinationTextarea.value = `Clipboard read failed: ${error}`;
  }
```

see https://github.com/mdn/content/pull/31785#discussion_r1457694159

File: files/en-us/web/api/clipboard/read/index.md [hunk 6]
```diff
@@ -240,6 +244,107 @@ Notes:
...
+    const clipboardContents = await navigator.clipboard.read({
+      unsanitized: ["text/html"],
+    });
+    const html = await getHTMLFromClipboardContents(clipboardContents);
+    if (html) {
+      destinationTextarea.value = html;
+    }
+  } catch (error) {
+    destinationTextarea.value = `Clipboard read failed: ${e}`;
+  }
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:338]:
```suggestion
  } catch (error) {
    destinationTextarea.value = `Clipboard read failed: ${error}`;
  }
```

see https://github.com/mdn/content/pull/31785#discussion_r1457694159

File: files/en-us/web/api/clipboard/read/index.md [hunk 7]
```diff
@@ -240,6 +244,107 @@ Notes:
...
+  const text = sourceTextarea.value;
+  const type = "text/html";
+  const blob = new Blob([text], { type });
+  const data = [new ClipboardItem({ [type]: blob })];
+
+  try {
+    await navigator.clipboard.write(data);
+  } catch (e) {
+    destinationTextarea.value = `Clipboard write failed: ${e}`;
+  }
```
Comment (skyclouds2001) [inline files/en-us/web/api/clipboard/read/index.md:298]:
```suggestion
  } catch (error) {
    destinationTextarea.value = `Clipboard write failed: ${error}`;
  }
```

for consistant, use `error` is better than `e`

also can use `e`

File: files/en-us/web/api/clipboard/read/index.md [hunk 8]
```diff
@@ -240,6 +244,107 @@ Notes:
 - If prompted, you will need to grant permission in order to paste the image.
 - This may not work on chromium browsers as the sample frame is not granted the [Permissions-Policy](/en-US/docs/Web/HTTP/Headers/Permissions-Policy) `clipboard-read` and `clipboard-write` permissions ([required by Chromium browsers](/en-US/docs/Web/API/Clipboard_API#security_considerations)).
 
+### Reading unsanitized HTML from the clipboard
+
+This example uses the `read(formats)` method to read HTML data from the clipboard and get the code in its original form, without the browser sanitizing it first.
+
+Note that, currently, only Chromium-based browsers support the `formats` parameter.
```
Comment (wbamberg) [inline files/en-us/web/api/clipboard/read/index.md:251]: I would not include this for a couple of reasons:
- keep compat in the compat section
- as a rule I don't think examples should "tell", they should only "show" - that is, they shouldn't tell you anything new, they should only illustrate things that the rest of the page has told you.

**Review by wbamberg — COMMENTED** [review]
Thank you @captainbrosset , this looks great. It's especially nice to have a good live sample.

> Apple and Mozilla might not actually ever implement this. Even though it's standardized and implemented in chromium, not all browsers are going to agree with it. Mozilla and Apple have said they want the behavior to be consistent whether you use the new async clipboard API or the old one, and that the HTML should either always be sanitized, or never, but that having an option makes no sense to them.
> 
> Mozilla: [mozilla/standards-positions#769 (comment)](https://github.com/mozilla/standards-positions/issues/769#issuecomment-1852250385) Apple: I couldn't find a standards position for webkit.

See also https://github.com/orgs/mdn/discussions/463 - we wanted to add this for the privacy sandbox APIs, but the Mozilla MDN team are opposed, so it's not happening, not yet anyway. So let's just not worry about representing it.

**Review by wbamberg — APPROVED** [review]
👍 thank you @captainbrosset !

**Pull request merged**